### DISCUSSION TITLE
Read from json (or plain text) with ToonzScript

### DISCUSSION THREAD

**Discussion by bencresty** [opentoonz/opentoonz#6375] [Ideas]
### Describe the solution you'd like

I want to import data from json to quickly create/setup/rig scenes in OpenToonz, but to my big surprise it doesn't seem to be possible to read json, xml or even plain text files. I searched all docs I could find about ToonzScript and tried some things, but didn't find anything about it. IMO functions to import and export data are very important features needed for many use cases when using a script language to automate work.
This surprises me especially because ToonzScript is pretty much like Javascript, so I wouldn't expect it to be very hard to add a module to import and/or export from/to json as there's so much out there already made. But perhaps I'm missing something.

Please! Add a way to import and export from/to json (especially import!)! That way the use case of ToonzScript expands hugely as we can develop our own tools in for instance NodeJs and export resulted data to json for OpenToonz to import with our own import scripts. 

Use case I need it for at this time: Importing spine.json files (rigging) as created with Affinity Designer. So that we don't need to setup the full rigging of a character again in OpenToonz as it's already automatically done in Affinity Designer.

Thanks in advance!

**Comment by RodneyBaker** [comment]
The primary problem here is likely toonzscript itself which needs (ideally) a total rewrite and whole new paradigm for scripting focused on external scripting not tied directly into the opentoonz core code.

I'd like to say that we could reasonably expect something like this by version 2.0 but since no developer has stated interest in this it would seem to be more of a target for v2.5 and v3.0.  In other words, a very long way away.
This perhaps especially given your stated interest being rigging of characters.

At the end of your request you mention a more likely approach than use of toonzscript and that would be import of files from Affinity Designer (and presumably other similar programs).  In a number of places Opentoonz itself does read in text files, xml and json.  That can perhaps be best seen in places such as XDTS import/export, OCA export (currently no import though) and others.  Various means of importing text files (of various formatting) can be seen via Function Editor where channel data can be saved out (and opened) as .curve files and data dumps.  The latter here being useful for use with File Interpolation where we point OT to a file that in turn supplies channel data.  This and more of course... to include most of Opentoonz files which tend to be in XML format such as the scene files (.tnz) themselves which would hold rigging related data.

If you are a scripter then my suggestion in the short term would be to approach from external scripting with an eye to identify any shortfalls that prevent that external access.  Most of that will relate to the Qt code as that is what Opentoonz is likely using to facilitate that process.  

Speaking of Qt...
This points to another potential future fix when Opentoonz eventually can update Qt to a version that better supports requirements.  A problem here is that Qt itself is a moving target that likely cannot be resolved in the near term.  This suggest to me that specific Qt related issue likely have a timeframe for resolution at Opentoonz v.2 and beyond.  In other words, potentially three or more years out.

**Comment by bencresty** [comment]
Thanks for the extensive answer @RodneyBaker

**Comment by RodneyBaker** [comment]
This request (one of the oldest feature requests submitted to Opentoonz) deals more with exporting but I am including it here for completeness sake:  opentoonz/opentoonz_docs#152

The primary difference (and therefore focus of interest) for this request appears to concern toonzscript.
It will be good to peel back the layers of where toonzscript currently is versus where users require it to be.

At present there appears to be no user/studio) actively using toonzscript although there are a few working demonstrations of toonzscript (mostly via youtube) dating back to when Opentoonz was first released.  These demonstrations appears very limited.  Perhaps the best example might be by Simon:  

https://www.youtube.com/watch?v=dYQCcn7WosM

Note that in the demonstration he states toonscript is 'relatively limited'.

The older Toonz website has some scripts but I personally have not been able to run them all successfully.

Of note:  Several enhancements to scripting have been added to Opentoonz since release to include capability to run scripts from command line more effectively.  Most of these relate in some way to development requested by and supported through the Morevna fork of Opentoonz.

**Comment by RodneyBaker** [comment]
@bencresty 
Something that will assist greatly wth this request would be to share examples of:  1) scripts that could/would likely be used 2) desired processes and user interactivity preferences  3) actual desired results/output

I specifically mention these because they can help in implementing the final solution but also pave ways to interim solutions (presumably outside of toonzscript) that achieve the same or similar results.

In other words, if toonzscript is not a near term solution... what other avenues of success can we pursue to meet your requirements?

This helps give us some solid targets to hit.

**Comment by bencresty** [comment]
> At present there appears to be no user/studio) actively using toonzscript although there are a few working demonstrations of toonzscript (mostly via youtube) dating back to when Opentoonz was first released. These demonstrations appears very limited. 

> The older Toonz website has some scripts but I personally have not been able to run them all successfully.

Without even knowing their reasons I can already tell what will most probably the reason; uses cases for scripting are extremely slim when developers aren't able to import or export from and to files. IMO scripting in software like this is used for these three reasons:
- Automation of frequently executed tasks (doing batch-stuff and firing macros)
- Use external data (import and process by code) to build scenes, animations etc. / create custom importers from data to be able to use other systems' outputs or custom data-files created by other custom software for all kinds of reasons
- Export data (could be images as well) to (data-)file(s). 

For this reason most modern software have each and every function and variable/object used inside the software available to be used within the scripting language. That's needed when using scripts, because the point of having a scripting language is to do custom things and you never know what a customer wants to use upfront. Next to this normally a scripting language has:
- io (so import/export from/to different types of files, mainly text and binary, but preferrably json and in this case xml because it's used so often in OpenToonz)
- math functions
- access to timing functions, a time object and the current time
- different types of variables (integer, float, strings etc.) and arrays of them
- conditionals (if.. else, if... then... else, switch ()... case)
- loops (for ... do, while ... do, for each... do)
- nowadays something like a mapping function
- things I'm missing now :)
- etc.

> Of note: Several enhancements to scripting have been added to Opentoonz since release to include capability to run scripts from command line more effectively. Most of these relate in some way to development requested by and supported through the Morevna fork of Opentoonz.

Great to know. I wonder tho, with the knowledge we have now that OpenToonz is very limited, if it wouldn't be faster to replace the whole thing by an already existing implementation of Javascript. That already contains all necesary stuff and I'd say it's "only" a matter of connecting the dots to the internal system. I can't tell how difficult that is in OpenToonz, but I get the feeling that's easier to do than to change this OpenToonz scripting language, which is basically incomplete and, as you said, not even really used. But that's up to you of course.

> @bencresty Something that will assist greatly wth this request would be to share examples of: 1) scripts that could/would likely be used 2) desired processes and user interactivity preferences 3) actual desired results/output

Sure I could send you things, but that doesn't help anything scripting, right? Than you're still developing something custom hardcoded in the software instead of letting users build it's own features. Also, this time it would be a Spine import, another time I'd like to import other data which will be completely different. The whole point of having a scripting language is to create custom code that can reach the inner workings of the software and connect that to other tools and/or data. Or to automate things, where you also would like to have a way to store preferences, variables, data etc. when doing it right and making things scalable.

I think it's much more worth you guys time to implement a solid scripting engine than to implement all kinds of user requests for inputs and outputs of data and whatever more people think of. That will continue a life time as there will always be new systems.

Also when implementing a solid and complete scripting system you make it possible for users to create modules or even plugins with an interface in OpenToonz, that users can share among each other to make OpenToonz even greater to specific needs, just by installing plugins or using custom scripts made by others. That's also a reason why many graphical and animation software these days decide to quickly build their scripting system, because it lifts their product up in a blink. Because people start making all kinds of scripts useful to others. Sometimes even to create things the developers of the software themselves never thought about or create temporary workarounds for things that have issues.  

So having a good scripting engine isn't only good for the advanced user who knows how to use it, it can also be very benificial to users that only know how to install or use a pre-made script. AND to you, because it can also be a great way to release the presure to have all feature requests added immediately (if the scripting language is good there and preferrably there will be some plugin system, than users create their own added functionality so you/opentoonz developers don't have to figure all that stuff out). And it happens often in other software that a custom made script is so good that thought out and used by many people that it ends up being included in the software, whether it's by rebuilding internally to act the same way, or to include the external script in the install package.

Hope you see the benefits and this makes sense.

BTW in another thread you were talking about 'us users', but you seem very active and the only one (?) reacting here to user feedback AFAIK and I believe you wrote somewhere you're not a developer (please correct me if I'm wrong, than sorry); what is your function in all this?

**Comment by RodneyBaker** [comment]
> BTW in another thread you were talking about 'us users', but you seem very active and the only one (?) reacting here to user feedback AFAIK and I believe you wrote somewhere you're not a developer (please correct me if I'm wrong, than sorry); what is your function in all this?

That is an excellent question and even after a years time I still don't have an adequate answer.

**Comment by RodneyBaker** [comment]
I have stared at this 'problem' of getting toonzscript to read json and plain text and I think the best I can come up with at the moment is to have a python helper script that actually alters the .toonzscript by placing the contents of the text file into the script.

Let's say we have a simple python script that copies all the contents of a text file into the toonscript as variables:  

```
# read_text_to_js.py
from pathlib import Path

BASE = Path(r"C:\OpenToonz stuff\sandbox\scripts")
txt_path = BASE / "example.txt"
out_path = BASE / "toon_data.js"

text = txt_path.read_text(encoding="utf-8")

def escape_js_string(s):
    return (
        s.replace("\\", "\\\\")
         .replace("\"", "\\\"")
         .replace("\r", "")
         .replace("\n", "\\n")
    )

escaped = escape_js_string(text)

out_path.write_text('var ExternalText = "{}";\n'.format(escaped), encoding="utf-8")
print("Wrote", out_path)
```

Basic example toonzscript that counts words:
```
function buildWordCounts(text) {
    var words = getCleanWords(text);
    var counts = {};   // plain JS object as dictionary

    for (var i = 0; i < words.length; i++) {
        var w = words[i];
        if (!w) continue;
        if (!counts[w]) counts[w] = 0;
        counts[w]++;
    }
    return counts;
}

var wordCounts = buildWordCounts(ExternalText);

// Print a few entries
var printed = 0;
for (var key in wordCounts) {
    if (!wordCounts.hasOwnProperty(key)) continue;
    print(key + " : " + wordCounts[key]);
    printed++;
    if (printed >= 20) break;
}
```



A simple toonzscript to capture that text: 

```
run("C:\\OpenToonz stuff\\sandbox\\scripts\\toon_data.js");
// ExternalText is now our entire file as a string

// Do *all* parsing in ToonzScript:
var words = getCleanWords(ExternalText);
// …whatever you want to do with words…
```
There may be limitations on just how long our text string can be.

**Comment by RodneyBaker** [comment]
Actually, here's a generic output from a working example where we use a .js file to store the text (instead of a json or standard text file).  We then populate our string variable based on that helper file:

```
run("C:\\OpenToonz stuff\\sandbox\\scripts\\read_data_demo.js")
Using ExternalText from toonz_data.js
---- Raw text ----
Hello from toonz_data.js!
This text lives in an external file.
Word count: 10
Word 0: hello
Word 1: from
Word 2: toon_data.js
Word 3: this
Word 4: text
Word 5: lives
Word 6: in
Word 7: an
Word 8: external
Word 9: file.
Done.
```
